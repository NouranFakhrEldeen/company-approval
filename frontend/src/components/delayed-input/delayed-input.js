import React, { forwardRef, useState, useEffect } from 'react';
import './delayed-input.style.scss';
const delayedInput = forwardRef(({
  onChange,
  name,
  className,
  placeholder,
  placeholderOverlay,
  value,
  delay,
  disabled,
  type,
  icon,
}, ref) => {
  const [data, setData] = useState({ target: { value: value || '', name } });
  const [val, setVal] = useState('');
  const debouncedTerm = useDebounce(data || {}, delay);
  useEffect(() => {
    onChange(debouncedTerm);
  }, [debouncedTerm]);

  const renderPlaceholder = (message, required, ref)=>{
    return (
      <div className={`placeholder ${icon ? 'with-icon' : ''}`} onClick={()=>ref.current?.focus()}>
        <label>
          {message}
        </label>
        {required ?
          <span className='fas fa-asterisk required'/> : null}
      </div>
    );
  };
  useEffect(() => {
    value && setVal(value);
  }, [value]);
  return (
    <div className={`delayed-input-component margin-bottom-20 ${className || ''}`}>
      { icon ? <span className="input-icon"><span className={`icon ${icon}`} /></span> : null}
      <div className={`input-container ${icon ? 'with-icon' : ''}`}>
        {!val?.length && placeholderOverlay ?
          renderPlaceholder(placeholderOverlay.message, placeholderOverlay.required, ref) : null }
        <textarea
          className={`${icon ? 'with-icon' : ''}`}
          disabled={disabled}
          placeholder={placeholder}
          ref={ref}
          value={val}
          onChange={e => {setData(e); setVal(e.target.value);}}
          type={type}
          name={name}
          data-testid={'delayed-input-it'}
          rows="1"
        />
      </div>
    </div>
  );
});

export const DelayedInput = delayedInput;

function useDebounce(value, delay) {
  const [debouncedValue, setDebouncedValue] = useState(value);
  useEffect(() => {
    const handler = setTimeout(() => setDebouncedValue(value), delay);
    return () => clearTimeout(handler);
  }, [value]);
  return debouncedValue;
}