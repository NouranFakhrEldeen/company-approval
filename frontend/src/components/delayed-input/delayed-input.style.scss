@import '../../styles/variables.scss';

.delayed-input-component {
  display: inline-block;
  position: relative;
  min-height: 35px;
  .placeholder {
    position: absolute;
    top: 25%;
    left: 12px;
    font-size: 14px;
    color: #77787a;
    width: 100%;
    label {
      max-width: calc(100% - 35px);
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }
    .required {
      font-size: 12px;
      position: relative;
      top: -13px;
      color: #ff741a;
    }
  }
  .placeholder.with-icon {
    left: 47px;
  }
  textarea {
    min-height: 35px;
    height: 100%;
    width: 100%;
    border: 1px solid #ccc;
  }
  .input-container.with-icon {
    width: calc(100% - 35px);
  }
  .input-container {
    min-height: 35px;
    width: 100%;
    height: 100%;
    display: inline-block;
    padding: 0;
  }
  .input-icon {
    min-height: 35px;
    height: 100%;
    background-color: #ff741a;
    width: 35px;
    display: inline-block;
    color: $white-color;
    text-align: center;
    font-size: $input-icons-size;
    padding: 6px;
  }
}
