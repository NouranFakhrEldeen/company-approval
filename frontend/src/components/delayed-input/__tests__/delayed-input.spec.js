import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
// import { HashRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { DelayedInput } from '../delayed-input';

let onChange = jest.fn();

describe('testing delayed input', ()=>{
  afterEach(cleanup);


  test('should checkbox checked', () => {
    const { getByTestId } = render(<DelayedInput data-testid='delayed-input-it' onChange={()=> onChange()}/>);
    const input = getByTestId('delayed-input-it');
    fireEvent.change(input, { target: { value: '23' } });
    expect(input.value).toBe('23');
  });

  //   snapshot
  test('should do snapshot', () => {
    const datePicker = renderer.create(
      <DelayedInput
        onChange={()=> onChange()}
      />
    ).toJSON();
    expect(datePicker).toMatchSnapshot();
  });

});