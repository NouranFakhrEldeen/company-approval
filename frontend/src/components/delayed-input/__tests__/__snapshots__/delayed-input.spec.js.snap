// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing delayed input should do snapshot 1`] = `
<div
  className="delayed-input-component margin-bottom-20 "
>
  <div
    className="input-container "
  >
    <textarea
      className=""
      data-testid="delayed-input-it"
      onChange={[Function]}
      rows="1"
      value=""
    />
  </div>
</div>
`;
